test/files/pos/t12312.scala

```diff
@@ -0,0 +1,8 @@
+class A { object X }
+
+class C {
+  val a = new A; import a.X
+  class D {
+    def res1() = X.isInstanceOf[X.type]
+  }
```

Comment:
Worth converting this to a run test to check that the semantics of the `instanceOf` check are correct. Maybe its covered by another test, but not necessarily.

Also worth testing something with multiple outer-pointer hops.

```scala
class A { object X }

object Test {
  val a, b = new A; import a.X
  class D {
    def isInstanceOf_aX(z: AnyRef) = z.isInstanceOf[X.type]
    class E {
      def isInstanceOf_aX(z: AnyRef) = z.isInstanceOf[X.type]
    }
  }
  def main(args: Array[String]): Unit = {
    val d = new D()
    assert(d.isInstanceOf_aX(a.X))
    assert(!d.isInstanceOf_aX(b.X))
    assert(!d.isInstanceOf_aX(new Object))

    val e = new d.E()
    assert(e.isInstanceOf_aX(a.X))
    assert(!e.isInstanceOf_aX(b.X))
    assert(!e.isInstanceOf_aX(new Object))
  }
}
```

The doubly nested case seems gives a `ClassCastException` as this PR stands:

```
Welcome to Scala 2.13.6-20210419-144201-bb7bea5 (OpenJDK 64-Bit Server VM, Java 1.8.0_272).
Type in expressions for evaluation. Or try :help.

...

scala> Test.main(null)
java.lang.ClassCastException: Test$D cannot be cast to Test$
  at Test$D$E.isInstanceOf_aX(<console>:8)
  at Test$.main(<console>:18)
  ... 32 elided
```